Add support for Duration & Distance in Table API.

mjkillough opened the pull request:
Add the ability to specify whether the table should be annotated with
distance and/or duration, as well as a new API for returning the
distance. This is done in a way that allows us to return errors for
unrouteable routes (where `null` is returned in the JSON), as in #7.

I'm not 100% sure of the use of `stdbool` in a FFI interface - it's been
a while since I had to write C!

Commit:
Extract error information from JSON.

When OSRM returns an error, it (often?) includes an error code/message
in the returned JSON object. This change extracts this code/message and
returns it to the caller of libosrmc. This makes it easier to debug
failures, but also makes it possible to react to errors such as
`NoRoute`.

The errors are extracted as both `code`/`message`, and a new
`osrmc_error_code` is introduced to allow the caller to extract the
code.

mjkillough commented on libosrmc/osrmc.cc:
```diff
@@ -245,14 +256,40 @@ float osrmc_table_response_duration(osrmc_table_response_t response, unsigned lo
+  auto optional = durations_from_to_all.values.at(to);
+
+  if (optional.is<osrm::json::Null>()) {
+    *error = new osrmc_error{"NoRoute: Impossible route between points"};
```
This is the same error that the Route API returns, which makes it easier for us to match on the string in the application. I'm very open to more robust suggestions as well!

daniel-j-h replied:
Returning an error and infinity is fine I think.

But we should document especially the infinity behavior and make it clear to the user that this is sort of a soft error.

The user does not have to shut it down just because there is no route between two sources in a huge table. Instead the user could e.g. use a great circle distance approximation in this case.

mjkillough replied:
Great! I'll actually be doing something very similar if I see an `INFINITY`. :-)

I've added a comment to the header file.

daniel-j-h commented on libosrmc/osrmc.h:
```diff
@@ -1,3 +1,5 @@
+#include <stdbool.h>
```
stdbool is :+1: C99 guarantees for this header. And if you need pre-C99 you are sorta on your own anyway.

daniel-j-h commented on libosrmc/osrmc.h:
```diff
@@ -195,13 +197,17 @@ OSRMC_API float osrmc_route_response_duration(osrmc_route_response_t response, o
 
 OSRMC_API osrmc_table_params_t osrmc_table_params_construct(osrmc_error_t* error);
 OSRMC_API void osrmc_table_params_destruct(osrmc_table_params_t params);
+OSRMC_API void osrmc_table_params_set_annotations(osrmc_table_params_t params, bool return_duration, bool return_distance);
```
Do you think we should make the annotations API open for extension from the start? That is:
- [ ] creating a new type annotations, similar to e.g. the params
- [ ] providing functions to (for now only) set the distance flag in the annotations struct
- [ ] passing the annotations struct around in the apis here

Then later on we could easily extend the annotations if users have a need for it. As it is right now it will be hard and ugly on the API to extend it later on.

mjkillough replied:
> Do you think we should make the annotations API open for extension from the start? That is:
> 
>  * [x]  creating a new type annotations, similar to e.g. the params
>  * [x]  providing functions to (for now only) set the distance flag in the annotations struct
>  * [x]  passing the annotations struct around in the apis here
> 
> Then later on we could easily extend the annotations if users have a need for it. As it is right now it will be hard and ugly on the API to extend it later on.

Sure! I've done this and defaulted it to include `AnnotationsType::Duration`, with no way to remove this. I am not sure whether OSRM even allows this!

mjkillough replied:
BTW, in doing this, I have been copying the `AnnontationsType` around, as I don't want to make the lifetime of the `osrmc_table_params_t` depend on the lifetime of the `osrmc_table_annotations_t`.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -245,14 +256,40 @@ float osrmc_table_response_duration(osrmc_table_response_t response, unsigned lo
+  *error = new osrmc_error{e.what()};
+  return INFINITY;
+}
+
```
What happens if the user calls this function when the distance annotations were not set? Then the JSON "distances" lookup will probably fail throwing an exception? I think we should instead assert this, since it's a user code error calling the distance function without setting distance annotations.

mjkillough replied:
I've updated it to assert, but it's a bit ugly - we need to keep a copy of the annotations around in order to assert. I'd love for `libosrm` to return something more structured than JSON, and to push this error handling in there.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -245,14 +256,40 @@ float osrmc_table_response_duration(osrmc_table_response_t response, unsigned lo
-  auto duration = durations_from_to_all.values.at(to).get<osrm::json::Number>().value;
+  auto optional = durations_from_to_all.values.at(to);
+
+  if (optional.is<osrm::json::Null>()) {
```
For the record this would also resolve https://github.com/daniel-j-h/libosrmc/issues/1

mjkillough commented:
In re-working this, I wanted to depend on the error handling introduced #7. The changes for this PR are in their own commit, bcb78d1f5670637f9c23b1a6987c0b47df177b53.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -26,9 +26,32 @@ int osrmc_is_abi_compatible(void) { return osrmc_get_version() >> 16u == OSRMC_V
+}
+
+static void osrmc_error_from_json(osrm::json::Object* json, osrmc_error_t* error) try {
+  if (json == nullptr) {
```
Looks like you cherry-picked the other pull rerquest's commit. Just make sure to update this branch when you make changes (such as asserting not null here).

mjkillough replied:
I've been rebasing this on top of #7, so should hopefully avoid causing any conflicts.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -188,15 +211,34 @@ float osrmc_route_response_duration(osrmc_route_response_t response, osrmc_error
+  delete reinterpret_cast<osrm::TableParameters::AnnotationsType*>(annotations);
+}
+
+void osrmc_table_annotations_set_distance(osrmc_table_annotations_t annotations, osrmc_error_t* error) try {
```
Do you think we should provide a way to set and unset a specific annotation? As in an API like `enable_distance(true)` or `enable_distance(false)` for setting and unsetting the distance annotation.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -208,46 +250,86 @@ void osrmc_table_params_add_source(osrmc_table_params_t params, size_t index, os
 
+struct osrmc_table_response {
+  osrm::json::Object json;
+  osrm::TableParameters::AnnotationsType annotations;
```
Why do we store the annotations in the response object here? So that we can pass the annotations along since we don"t have a global `Table` object?

mjkillough replied:
I really wanted to avoid this, but at the time it was the only way I could see to produce an error if someone tried to get a distance from a table response that hadn't requested the distance. 

I think I was having a bad day yesterday, as I see now I don't need to check for this explicitly: we can [check whether the key is set in the response object](https://github.com/Project-OSRM/osrm-backend/blob/master/include/engine/api/table_api.hpp#L78-L88). I'll switch to doing that, and get rid of this extra `osrmc_table_response` struct.

daniel-j-h replied:
Ah, I didn't think of this, too! Good catch, let's do it this way then!

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -208,46 +250,86 @@ void osrmc_table_params_add_source(osrmc_table_params_t params, size_t index, os
+  if (!(response->annotations & osrm::TableParameters::AnnotationsType::Distance)) {
+    *error = new osrmc_error{"NoTable", "Table request not configured to return distances"};
+    return INFINITY;
+  }
```
This is the only use-case which requires us to pass the annotations object along in the response? It's a bit ugly.. maybe we can re-work this? If not I think it's okay for now.

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -208,46 +250,86 @@ void osrmc_table_params_add_source(osrmc_table_params_t params, size_t index, os
+  auto& distances_from_to_all = distances.values.at(from).get<osrm::json::Array>();
+  auto optional = distances_from_to_all.values.at(to);
+
+  if (optional.is<osrm::json::Null>()) {
```
optional is a C++ type, let's prevent potential future name clashes and rename this var maybe?

daniel-j-h reviewed (commented):
I left some comments inline, in general I feel like you are going in the right direction. Thanks for this! :sparkles:

daniel-j-h commented on bindings/osrm_c89.c:
```diff
@@ -84,7 +84,7 @@ int main(int argc, char** argv) {
   /* If we got here on a failure path notify the user */
   if (error) {
-    fprintf(stderr, "Error: %s\n", osrmc_error_message(error));
+    fprintf(stderr, "Error: code=%s, message=%s\n", osrmc_error_code(error), osrmc_error_message(error));
```
Seeing this change in the example bindings here: have you thought about ABI/API testing `libosrm -> libosrmc -> your rust bindings`? Right now this library doesn't really come with e.g. CI/CD or integration tests. Working on the C bindings might break the examples and/or your rust bindings. Just some food for thought.

mjkillough replied:
There are tests for the Rust bindings, but they're unfortunately not running in publicly visible CI right now. 

I consider the `libosrmc` bindings an implementation detail of the Rust bindings. A specific version is included as a git submodule, and it's then statically compiled into the Rust library. Thankfully we don't need to worry about working with arbitrary `libosrmc`s!

The `libosrm` is a bit trickier: we test against the specific version that we need, but we don't test all available versions. I figured I'd cross that bridge if anyone else came to depend on the Rust bindings. 🙂

Commit:
Add support for Duration & Distance in Table API.

Add the ability to specify whether the table should be annotated with
distance and/or duration, as well as a new API for returning the
distance. This is done in a way that allows us to return errors for
unrouteable routes (where `null` is returned in the JSON), as in #7.

I'm not 100% sure of the use of `stdbool` in a FFI interface - it's been
a while since I had to write C!

mjkillough commented:
OK! Hopefully I've gotten this a little closer, but please do let me know if you'd like further changes.

There really is no rush for this. I really appreicate you taking the time to provide such detailed feedback, especially as this isn't something you're actively maintaining! ❤️

daniel-j-h commented on libosrmc/osrmc.cc:
```diff
@@ -26,9 +26,30 @@ int osrmc_is_abi_compatible(void) { return osrmc_get_version() >> 16u == OSRMC_V
+}
+
+static void osrmc_error_from_json(osrm::json::Object* json, osrmc_error_t* error) try {
+  assert(json != nullptr);
```
See https://github.com/daniel-j-h/libosrmc/pull/7#discussion_r266243991

daniel-j-h commented:
Looks good; there's the pointer vs. ref thingy but then I think we can merge this in.

Do you want me to merge this pull request in and then close the other one, since you are pulling in the other pull request's commit. Or merge the other pull request in first, then rebase this one on master, and then merge this one in?

mjkillough commented:
> Do you want me to merge this pull request in and then close the other one, since you are pulling in the other pull request's commit. Or merge the other pull request in first, then rebase this one on master, and then merge this one in?

I'm happy for you to close the other one and merge this one in. I think GitHub might actually be smart enough to close both once you merge this. 🙂
